api/src/main/java/org/apache/iceberg/types/TypeUtil.java

```diff
@@ -319,20 +319,45 @@ public static boolean isPromotionAllowed(Type from, Type.PrimitiveType to) {
    */
   public static void validateWriteSchema(Schema tableSchema, Schema writeSchema,
                                          Boolean checkNullability, Boolean checkOrdering) {
+    String errMsg = "Cannot write incompatible dataset to table with schema:";
+    checkSchemaCompatibility(errMsg, tableSchema, writeSchema, checkNullability, checkOrdering);
+  }
+
+  /**
+   * Validates whether the provided schema is compatible with the expected schema.
+   *
+   * @param context the schema context (e.g. row ID)
+   * @param expectedSchema the expected schema
+   * @param providedSchema the provided schema
+   * @param checkNullability whether to check field nullability
+   * @param checkOrdering whether to check field ordering
+   */
+  public static void validateSchema(String context, Schema expectedSchema, Schema providedSchema,
+                                    boolean checkNullability, boolean checkOrdering) {
+    String errMsg = String.format("Provided %s schema is incompatible with expected %s schema:", context, context);
+    checkSchemaCompatibility(errMsg, expectedSchema, providedSchema, checkNullability, checkOrdering);
+  }
+
+  private static void checkSchemaCompatibility(String errMsg, Schema schema, Schema providedSchema,
+                                               boolean checkNullability, boolean checkOrdering) {
     List<String> errors;
     if (checkNullability) {
-      errors = CheckCompatibility.writeCompatibilityErrors(tableSchema, writeSchema, checkOrdering);
+      errors = CheckCompatibility.writeCompatibilityErrors(schema, providedSchema, checkOrdering);
     } else {
-      errors = CheckCompatibility.typeCompatibilityErrors(tableSchema, writeSchema, checkOrdering);
+      errors = CheckCompatibility.typeCompatibilityErrors(schema, providedSchema, checkOrdering);
     }
 
     if (!errors.isEmpty()) {
       StringBuilder sb = new StringBuilder();
-      sb.append("Cannot write incompatible dataset to table with schema:\n")
-          .append(tableSchema)
-          .append("\nwrite schema:")
-          .append(writeSchema)
-          .append("\nProblems:");
+      sb.append(errMsg)
+          .append("\n")
+          .append(schema)
+          .append("\n")
+          .append("provided schema:")
+          .append("\n")
+          .append(providedSchema)
+          .append("\n")
+          .append("problems:");
```

Comment:
Nit: I prefer capitalizing these like they were before. It looks weird to not use sentence case.

Reply:
We capitalized only one before so it looked inconsistent. Made both start with capital letters now.